### PR TITLE
Enhance INC WOQ model loading & support Huggingface WOQ model loading

### PR DESCRIPTION
**Pull request by yuwenzho**
## Type of Change

feature
API changed or not: yes

## Description

Enhance INC WOQ model loading and support Huggingface WOQ model loading:
design from @xin3he:
image

## How has this PR been tested?

CI

## Dependency Change?

no

### CODE REVIEW

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 1]
```diff
@@ -38,14 +38,472 @@ def save(model, output_dir="./saved_results"):
...
+    model.eval()
+    return model
+
+
+def _load_hf_woq_model(pretrained_model_name_or_path, *model_args, **kwargs):
+    # check required package
+    try:
+        import transformers
+    except ImportError:
+        logger.error("`transformers` package is required for loading hugginface weight-only quantization model.")
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:137]: This exception handling will log the error and continue the execution. I guest the expected behavior is to attempt to install the required package automatically. If failed, raise an import error.

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 2]
```diff
@@ -38,14 +38,472 @@ def save(model, output_dir="./saved_results"):
...
+
+    try:
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:139]: ditto

**Review by yiliu30 — APPROVED** [review]
Left some comments, others are LGTM

File: neural_compressor/torch/quantization/load_entry.py [hunk 3]
```diff
@@ -25,43 +25,63 @@
...
+from neural_compressor.torch.utils import LoadFormat
 
 config_name_mapping = {
     FP8_QUANT: FP8Config,
 }
 
 
-def load(output_dir="./saved_results", model=None):
-    """The main entry of load for all algorithms.
+def load(model=None, checkpoint_dir="./saved_results", format="default", *hf_model_args, **hf_model_kwargs):
```
Comment (xin3he) [inline neural_compressor/torch/quantization/load_entry.py:35]: checkpoint_dir is always required and please set it as args
`def load(checkpoint_dir, model=None, format="default", *hf_model_args, **hf_model_kwargs)`
I don't expect to change all load behavior in this PR.

**Review by xin3he — APPROVED** [review]

**Review by xin3he — CHANGES_REQUESTED** [review]
please change the load behavior back.

File: neural_compressor/torch/quantization/load_entry.py (hunk 3 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/load_entry.py]: Updated to new design, now is `load(model_name_or_path, original_model=None, format='default', device='cpu', *model_args, **kwargs)`

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 4]
```diff
@@ -37,14 +38,516 @@ def save(model, output_dir="./saved_results"):
...
+            all remaining positional arguments for loading huggingface models.
+            will be passed to the huggingface model's `__init__` method.
+        kwargs (remaining dictionary of keyword arguments, optional):
+            remaining dictionary of keyword arguments for loading huggingface models.
+            will be passed to the huggingface model's `__init__` method, such as 'trust_remote_code', 'revision'.
+    Returns:
+        torch.nn.Module: quantized model
+    """
+    if format == LoadFormat.HUGGINGFACE:
+        model = _load_hf_woq_model(model_name_or_path, *model_args, **kwargs)
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:74]: The loaded model is still INC provided model, how about using `_load_hf_format_model` and `_load_inc_format_model`?

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 5]
```diff
@@ -37,14 +38,516 @@ def save(model, output_dir="./saved_results"):
...
+    with open(qconfig_file_path, "r") as file:
+        quantization_config = json.load(file)
+
+    model = _build_woq_model(origin_model, quantization_config, qweights.keys())
+    model.load_state_dict(qweights, assign=True)
+    model.eval()
+    return model
+
+
+def _load_hf_woq_model(pretrained_model_name_or_path, *model_args, **kwargs):
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:180]: how about splitting this function into several parts? we will convert format from autogptq to hpu in this function and currently this function is too long to understand.

File: neural_compressor/torch/quantization/load_entry.py [hunk 6]
```diff
@@ -25,43 +25,87 @@
...
+        device (str, optional): 'cpu', 'hpu' or 'cuda'. specify the device the model will be loaded to.
+        model_args (sequence of positional arguments, optional):
+            all remaining positional arguments for loading huggingface models.
+            Will be passed to the huggingface model's `__init__` method.
+        kwargs (remaining dictionary of keyword arguments, optional):
+            remaining dictionary of keyword arguments for loading huggingface models.
+            Will be passed to the huggingface model's `__init__` method, such as 'trust_remote_code', 'revision'.
     Returns:
-        The quantized model
+        torch.nn.Module: quantized model
```
Comment (xin3he) [inline neural_compressor/torch/quantization/load_entry.py:76]: not only nn.Module. includes TorchScript model with IPEX and fx graph with pt2e.

File: neural_compressor/torch/quantization/load_entry.py [hunk 7]
```diff
@@ -25,43 +25,87 @@
...
+            If 'format' is set to 'huggingface', it means the huggingface model_name_or_path.
+            If 'format' is set to 'default', it means the 'checkpoint_dir'.
+            Parameter should not be None. it coworks with 'original_model' parameter to load INC
+            quantized model in local.
+        original_model (torch.nn.module, optional): original model before quantization.
+            Needed if 'format' is set to 'default' and not TorchScript model.Defaults to None.
+        format (str, optional): 'defult' for loading INC quantized model.
+            'huggingface' for loading huggingface WOQ causal language model. Defaults to "default".
+        device (str, optional): 'cpu', 'hpu' or 'cuda'. specify the device the model will be loaded to.
+        model_args (sequence of positional arguments, optional):
```
Comment (xin3he) [inline neural_compressor/torch/quantization/load_entry.py:69]: the same as above

**Review by xin3he — APPROVED** [review]

File: neural_compressor/torch/quantization/load_entry.py (hunk 7 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/load_entry.py]: updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68

File: neural_compressor/torch/quantization/load_entry.py (hunk 6 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/load_entry.py]: updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68

File: neural_compressor/torch/algorithms/weight_only/save_load.py (hunk 5 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68

File: neural_compressor/torch/algorithms/weight_only/save_load.py (hunk 4 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 8]
```diff
@@ -37,14 +38,516 @@ def save(model, output_dir="./saved_results"):
...
+        model_name_or_path (str):  torch checkpoint directory or hugginface model_name_or_path.
+            If 'format' is set to 'huggingface', it means the huggingface model_name_or_path.
+            If 'format' is set to 'default', it means the 'checkpoint_dir'.
+            Parameter should not be None. it coworks with 'original_model' parameter to load INC
+            weight-only quantized model in local.
+        original_model (torch.nn.module, optional): original model before quantization.
+            Needed if 'format' is set to 'default' and not TorchScript model.Defaults to None.
+        format (str, optional): 'defult' for loading INC weight-only quantized model.
+            'huggingface' for loading huggingface WOQ causal language model. Defaults to "default".
+        model_args (sequence of positional arguments, optional):
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 9]
```diff
@@ -37,14 +39,596 @@ def save(model, output_dir="./saved_results"):
...
+        self.format = format
+        self.device = device
+        self.kwargs = kwargs
+        self.quantization_config = {}
+        self.loaded_state_dict_keys = {}
+
+    def load_woq_model(self):
+        if self.format == LoadFormat.HUGGINGFACE:
+            model = self.load_hf_format_woq_model()
+            logger.info("Quantized huggingface model loading successful.")
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:90]: It seems we just load the `state_dicit` of the quantized model without doing quantization. If so, please update the log message.

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 10]
```diff
@@ -37,14 +39,596 @@ def save(model, output_dir="./saved_results"):
...
+        elif self.format == LoadFormat.DEFAULT:
+            qmodel_weight_file_path = os.path.join(
+                os.path.abspath(os.path.expanduser(self.model_name_or_path)), WEIGHT_NAME
+            )
+            assert os.path.exists(qmodel_weight_file_path), "Cannot load model weight from path {}".format(
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:95]: It only checks the existence of the file but does not make sure it's a qmodel weight.
Suggest raising a message to guide users on which values should be passed if they want to load the qmodel with `DEFAULT` format.

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 11]
```diff
@@ -37,14 +39,596 @@ def save(model, output_dir="./saved_results"):
...
+                    zp=zp,
+                    bias=module.bias is not None,
+                    use_optimum_format=True,
+                    **kwargs,
+                )
+                set_module(self.original_model, name, new_module)
+        woq_model = self.original_model
+        return woq_model
+
+    def _check_required_packages(self):
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/save_load.py:220]: It should be a static method or util function.

File: neural_compressor/torch/quantization/load_entry.py [hunk 12]
```diff
@@ -25,43 +25,85 @@
...
 
-def load(output_dir="./saved_results", model=None):
-    """The main entry of load for all algorithms.
+def load(model_name_or_path, original_model=None, format="default", device="cpu", **kwargs):
+    """Load quantized model.
 
-    Args:
-        output_dir (str, optional): path to quantized model folder. Defaults to "./saved_results".
-        model (torch.nn.Module, optional): original model, suggest to use empty tensor.
+    1. Load INC quantized model in local.
```
Comment (yiliu30) [inline neural_compressor/torch/quantization/load_entry.py:38]: nit: Please align the introduction and cases. For example:
1. INC 
      case 1: WOQ + local
      case 2: INT8/FP8 + local
      ...
2. HF
      ...

File: neural_compressor/torch/quantization/load_entry.py [hunk 13]
```diff
@@ -25,43 +25,85 @@
...
 
-            return load(output_dir)
+                return load(model_name_or_path, original_model, format=LoadFormat.DEFAULT)
 
-        model.qconfig = config_mapping
-        if isinstance(config_object, FP8Config):  # FP8
-            from neural_compressor.torch.algorithms.habana_fp8 import load
+            original_model.qconfig = config_mapping
+            if isinstance(config_object, FP8Config):  # FP8
+                from neural_compressor.torch.algorithms.habana_fp8 import load
```
Comment (yiliu30) [inline neural_compressor/torch/quantization/load_entry.py:100]: nit: there are several `load` functions, maybe using the Google style imports makes it easier to understand.

```python
import neural_compressor.torch.algorithms.habana_fp8 as fp8
fp8.load(...)
```

**Review by yiliu30 — APPROVED** [review]

**Comment by yiliu30** [comment]
@xin3he The load API is one of the important user-facing API, how about creating a discussion topic to track the design process.

**Review by Kaihui-intel — APPROVED** [review]

File: examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/habana_fp8/run_llm.py [hunk 14]
```diff
@@ -63,7 +63,7 @@
 parser.add_argument("--calib_iters", default=100, type=int,
                     help="calibration iters.")
 parser.add_argument("--tasks", nargs='+', default=["lambada_openai"], \
-                    type=str, choices=["hellaswag", "lambada_openai", "piqa", "winogrande", "copa", 
+                    type=str, choices=["hellaswag", "lambada_openai", "piqa", "winogrande", "copa",
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/habana_fp8/run_llm.py:66]: The habana_fp8 example will be removed once we sync from gerrit INC-folk which uses HQT FP8. So I will not modify it here.

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 15]
```diff
@@ -37,14 +39,596 @@ def save(model, output_dir="./saved_results"):
...
+        try:
+            import accelerate
+            import transformers
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: update in https://github.com/intel/neural-compressor/pull/1826/commits/42cfe39722bad7366a66ff3701e7d3e68e0e31e8: use `is_package_available` to check package is available or not.

File: neural_compressor/torch/quantization/load_entry.py (hunk 13 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/load_entry.py]: update in https://github.com/intel/neural-compressor/pull/1826/commits/a33c766e8865f009ab02fa4ccd24cedd85574113

File: neural_compressor/torch/quantization/load_entry.py (hunk 12 above)
Comment (yuwenzho) [reply neural_compressor/torch/quantization/load_entry.py:38]: update in https://github.com/intel/neural-compressor/commit/a33c766e8865f009ab02fa4ccd24cedd85574113

File: neural_compressor/torch/algorithms/weight_only/save_load.py (hunk 11 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: update in https://github.com/intel/neural-compressor/commit/42cfe39722bad7366a66ff3701e7d3e68e0e31e8: use is_package_available to check package is available or not.

File: neural_compressor/torch/algorithms/weight_only/save_load.py (hunk 10 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: add log info in https://github.com/intel/neural-compressor/commit/a33c766e8865f009ab02fa4ccd24cedd85574113

File: neural_compressor/torch/algorithms/weight_only/save_load.py (hunk 9 above)
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: update log info in https://github.com/intel/neural-compressor/commit/a33c766e8865f009ab02fa4ccd24cedd85574113

File: neural_compressor/torch/algorithms/weight_only/save_load.py [hunk 16]
```diff
@@ -37,14 +39,596 @@ def save(model, output_dir="./saved_results"):
...
+            Needed if 'format' is set to 'default' and not TorchScript model.Defaults to None.
+        format (str, optional): 'defult' for loading INC weight-only quantized model.
+            'huggingface' for loading huggingface WOQ causal language model. Defaults to "default".
+        kwargs (remaining dictionary of keyword arguments, optional):
+            remaining dictionary of keyword arguments for loading huggingface models.
+            will be passed to the huggingface model's `__init__` method, such as 'trust_remote_code', 'revision'.
+    Returns:
+        torch.nn.Module: quantized model
+    """
+    model_loader = WOQModelLoader(model_name_or_path, original_model, format, device="cpu", **kwargs)
```
[the remark this answers is not in the document]
Comment (yuwenzho) [reply neural_compressor/torch/algorithms/weight_only/save_load.py]: update in https://github.com/intel/neural-compressor/commit/a33c766e8865f009ab02fa4ccd24cedd85574113

**Comment by yuwenzho** [comment]
> @xin3he The load API is one of the important user-facing API, how about creating a discussion topic to track the design process.

I added in the discussions of torch 3.x API https://github.com/intel/neural-compressor/discussions/1527

**Pull request merged**